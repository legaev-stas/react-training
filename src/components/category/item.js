import React from 'react';
import {SwipeAction, List, Modal, Badge} from 'antd-mobile';


export class CategoryItem extends React.Component {
    constructor(props) {
        super(props);

        this.editHandler = this.editHandler.bind(this);
        this.deleteHandler = this.deleteHandler.bind(this);
        this.openCategory = this.openCategory.bind(this);
    }

    deleteHandler() {
        let heading = 'Delete Category';
        let message = `Are you sure? The category contains ${this.props.uncompletedTasksNumber} uncompleted tasks`;
        if (this.props.uncompletedTasksNumber) {
            Modal.alert(heading, message, [
                {text: 'Cancel', style: 'default'},
                {text: 'OK', onPress: () => this.props.deleteHandler(this.props.id)},
            ]);
        } else {
            this.props.deleteHandler(this.props.id);
        }
    }

    openCategory() {
        this.props.openCategory(this.props.id);
    }

    editHandler() {
        let that = this;
        Modal.prompt('Update category name', '',
            [
                {text: 'Cancel'},
                {
                    text: 'Update',
                    onPress: name => new Promise((resolve) => {
                        that.props.editHandler({
                            id: that.props.id,
                            name
                        });
                        resolve();
                    }),
                },
            ], 'default', that.props.name);
    }

    render() {
        return (
            <SwipeAction
                style={{backgroundColor: 'gray'}}
                autoClose
                right={[
                    {
                        text: 'Delete',
                        onPress: this.deleteHandler,
                        style: {backgroundColor: 'red', color: 'white'},
                    }
                ]}
                left={[
                    {
                        text: 'Edit',
                        onPress: this.editHandler,
                        style: {backgroundColor: 'blue', color: 'white'},
                    }
                ]}
            >
                <List.Item
                    onClick={this.openCategory}
                    extra={<Badge text={this.props.uncompletedTasksNumber} overflowCount={10}/>}
                >{this.props.name}</List.Item>
            </SwipeAction>
        );
    }
}
