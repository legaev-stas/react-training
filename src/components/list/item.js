import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import Immutable from 'immutable';
import {SwipeAction, List, Badge, Checkbox, Flex} from 'antd-mobile';
import './styles.css';


export class ListItem extends React.PureComponent {
    propTypes = {
        model: PropTypes.instanceOf(Immutable.Map).isRequired,
        onDelete: PropTypes.func,
        onEdit: PropTypes.func,
        onClick: PropTypes.func,
        onStatusChange: PropTypes.func,
        arrow: PropTypes.bool,
        checkable: PropTypes.bool
    }

    componentDidMount(){
        ReactDOM.findDOMNode(this).addEventListener('touchmove', this.preventScrollWhenSwipe, false);
    }
    componentWillUnmount(){
        ReactDOM.findDOMNode(this).removeEventListener('touchmove', this.preventScrollWhenSwipe, false);
    }

    preventScrollWhenSwipe = (e) => {
        if(ReactDOM.findDOMNode(this).querySelector('.am-swipe-swiping')){
            e.preventDefault();
        }
    }

    onDelete = () => {
        this.props.onDelete && this.props.onDelete(this.props.model.toJS());
    }

    onClick = () => {
        this.props.onClick && this.props.onClick(this.props.model.toJS());
    }

    onEdit = () => {
        this.props.onEdit && this.props.onEdit(this.props.model.toJS());
    }

    onStatusChange = (e) => {
        e.stopPropagation();
        this.props.onStatusChange({
            id: this.props.model.get('id'),
            completed: !this.props.model.get('completed')
        });
    }

    render() {
        return (
            <SwipeAction
                style={{backgroundColor: 'gray'}}
                autoClose
                right={this.props.onDelete && [
                    {
                        text: 'Delete',
                        onPress: this.onDelete,
                        style: {backgroundColor: 'red', color: 'white'},
                    }
                ]}
                left={this.props.onEdit && [
                    {
                        text: 'Edit',
                        onPress: this.onEdit,
                        style: {backgroundColor: 'blue', color: 'white'},
                    }
                ]}
            >
                <List.Item
                    extra={<Badge text={this.props.model.get('badge')} overflowCount={10}/>}
                    arrow={this.props.arrow && "horizontal"}
                    onClick={this.onClick}
                >
                    <Flex>
                        {this.props.checkable &&
                        <Checkbox checked={this.props.model.get('completed')} onClick={this.onStatusChange}/>
                        }
                        <Flex.Item>{this.props.model.get('title')}</Flex.Item>
                    </Flex>
                </List.Item>
            </SwipeAction>
        );
    }
}
