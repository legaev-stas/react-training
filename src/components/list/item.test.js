import React from 'react';
import {fromJS} from 'immutable';

import {shallow, render, mount, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';

configure({adapter: new Adapter()});

import {ListItem} from './item';


let model;
beforeEach(()=>{
    model = fromJS({
        id: 'id',
        completed: false,
        badge: 0,
        title: 'Title'
    });
});

it('should render right (delete) swipe option when onDelete property is provided', () => {
    const tree = mount(<ListItem model={model} onDelete={jest.fn}/>, {mode: 'shallow'});
    expect(toJson(tree)).toMatchSnapshot();
});

it('should render left (edit) swipe option when onEdit property is provided', () => {
    const tree = mount(<ListItem model={model} onEdit={jest.fn}/>, {mode: 'shallow'});
    expect(toJson(tree)).toMatchSnapshot();
});

it('should set arrow="horizontal" on nested List.Item when arrow property is provided', () => {
    const tree = mount(<ListItem model={model} arrow/>, {mode: 'shallow'});
    expect(toJson(tree)).toMatchSnapshot();
});

it('should rendered nested Checkbox when checkable property is provided', () => {
    const tree = mount(<ListItem model={model} checkable/>, {mode: 'shallow'});
    expect(toJson(tree)).toMatchSnapshot();
});

it('onStatusChange callback is invoked with correct arguments when Checkbox is clicked', () => {
    const onStatusChange = jest.fn();
    const tree = mount(<ListItem model={model} onStatusChange={onStatusChange} checkable/>, {mode: 'shallow'});
    const eventMock = {
        stopPropagation() {
        }
    };
    tree.find('Checkbox').simulate('click', eventMock);
    expect(onStatusChange).toHaveBeenCalledWith({
        id: 'id',
        completed: true,
    })
});

it('onDelete method should invoke callback onDelete with correct argument (model)', () => {
    const onDelete = jest.fn();
    const tree = mount(<ListItem model={model} onDelete={onDelete}/>, {mode: 'shallow'});
    tree.instance().onDelete();
    expect(onDelete).toHaveBeenCalledWith(model.toJS());
});

it('onEdit method should invoke callback onEdit with correct argument (model)', () => {
    const onEdit = jest.fn();
    const tree = mount(<ListItem model={model} onEdit={onEdit}/>, {mode: 'shallow'});
    tree.instance().onEdit();
    expect(onEdit).toHaveBeenCalledWith(model.toJS());
});

it('onClick method should invoke callback onClick with correct argument (model)', () => {
    const onClick = jest.fn();
    const tree = mount(<ListItem model={model} onClick={onClick}/>, {mode: 'shallow'});
    tree.instance().onClick();
    expect(onClick).toHaveBeenCalledWith(model.toJS());
});


