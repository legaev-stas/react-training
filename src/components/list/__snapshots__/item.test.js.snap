// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render left (edit) swipe option when onEdit property is provided 1`] = `
<ListItem
  model={
    Immutable.Map {
      "id": "id",
      "completed": false,
      "badge": 0,
      "title": "Title",
    }
  }
  onEdit={[Function]}
>
  <SwipeAction
    autoClose={true}
    disabled={false}
    left={
      Array [
        Object {
          "onPress": [Function],
          "style": Object {
            "backgroundColor": "blue",
            "color": "white",
          },
          "text": "Edit",
        },
      ]
    }
    onClose={[Function]}
    onOpen={[Function]}
    prefixCls="am-swipe"
    right={Array []}
    style={
      Object {
        "backgroundColor": "gray",
      }
    }
  >
    <ListItem
      align="middle"
      error={false}
      extra={
        <Badge
          corner={false}
          dot={false}
          overflowCount={10}
          prefixCls="am-badge"
          size="small"
          text={0}
        />
      }
      multipleLine={false}
      onClick={[Function]}
      platform="ios"
      prefixCls="am-list"
      wrap={false}
    >
      <Flex
        align="center"
        prefixCls="am-flexbox"
      >
        <FlexItem
          prefixCls="am-flexbox"
        >
          Title
        </FlexItem>
      </Flex>
    </ListItem>
  </SwipeAction>
</ListItem>
`;

exports[`should render right (delete) swipe option when onDelete property is provided 1`] = `
<ListItem
  model={
    Immutable.Map {
      "id": "id",
      "completed": false,
      "badge": 0,
      "title": "Title",
    }
  }
  onDelete={[Function]}
>
  <SwipeAction
    autoClose={true}
    disabled={false}
    left={Array []}
    onClose={[Function]}
    onOpen={[Function]}
    prefixCls="am-swipe"
    right={
      Array [
        Object {
          "onPress": [Function],
          "style": Object {
            "backgroundColor": "red",
            "color": "white",
          },
          "text": "Delete",
        },
      ]
    }
    style={
      Object {
        "backgroundColor": "gray",
      }
    }
  >
    <ListItem
      align="middle"
      error={false}
      extra={
        <Badge
          corner={false}
          dot={false}
          overflowCount={10}
          prefixCls="am-badge"
          size="small"
          text={0}
        />
      }
      multipleLine={false}
      onClick={[Function]}
      platform="ios"
      prefixCls="am-list"
      wrap={false}
    >
      <Flex
        align="center"
        prefixCls="am-flexbox"
      >
        <FlexItem
          prefixCls="am-flexbox"
        >
          Title
        </FlexItem>
      </Flex>
    </ListItem>
  </SwipeAction>
</ListItem>
`;

exports[`should rendered nested Checkbox when checkable property is provided 1`] = `
<ListItem
  checkable={true}
  model={
    Immutable.Map {
      "id": "id",
      "completed": false,
      "badge": 0,
      "title": "Title",
    }
  }
>
  <SwipeAction
    autoClose={true}
    disabled={false}
    left={Array []}
    onClose={[Function]}
    onOpen={[Function]}
    prefixCls="am-swipe"
    right={Array []}
    style={
      Object {
        "backgroundColor": "gray",
      }
    }
  >
    <ListItem
      align="middle"
      error={false}
      extra={
        <Badge
          corner={false}
          dot={false}
          overflowCount={10}
          prefixCls="am-badge"
          size="small"
          text={0}
        />
      }
      multipleLine={false}
      onClick={[Function]}
      platform="ios"
      prefixCls="am-list"
      wrap={false}
    >
      <Flex
        align="center"
        prefixCls="am-flexbox"
      >
        <Checkbox
          checked={false}
          onClick={[Function]}
          prefixCls="am-checkbox"
          wrapLabel={true}
        />
        <FlexItem
          prefixCls="am-flexbox"
        >
          Title
        </FlexItem>
      </Flex>
    </ListItem>
  </SwipeAction>
</ListItem>
`;

exports[`should set arrow="horizontal" on nested List.Item when arrow property is provided 1`] = `
<ListItem
  arrow={true}
  model={
    Immutable.Map {
      "id": "id",
      "completed": false,
      "badge": 0,
      "title": "Title",
    }
  }
>
  <SwipeAction
    autoClose={true}
    disabled={false}
    left={Array []}
    onClose={[Function]}
    onOpen={[Function]}
    prefixCls="am-swipe"
    right={Array []}
    style={
      Object {
        "backgroundColor": "gray",
      }
    }
  >
    <ListItem
      align="middle"
      arrow="horizontal"
      error={false}
      extra={
        <Badge
          corner={false}
          dot={false}
          overflowCount={10}
          prefixCls="am-badge"
          size="small"
          text={0}
        />
      }
      multipleLine={false}
      onClick={[Function]}
      platform="ios"
      prefixCls="am-list"
      wrap={false}
    >
      <Flex
        align="center"
        prefixCls="am-flexbox"
      >
        <FlexItem
          prefixCls="am-flexbox"
        >
          Title
        </FlexItem>
      </Flex>
    </ListItem>
  </SwipeAction>
</ListItem>
`;
